sway_libs/src/merkle_proof/binary_merkle_proof.sw

```diff
@@ -0,0 +1,175 @@
+// TODO: Function definitions that use arrays should be updated to Vecs once
+// https://github.com/FuelLabs/fuels-rs/issues/353 is resolved
+
+library binary_merkle_proof;
+
+use std::{hash::sha256, revert::require};
+
+pub enum ProofError {
+    InvalidKey: (),
+    InvalidProofLength: (),
+}
+
+pub const LEAF = 0u8;
+pub const NODE = 1u8;
+
+/// Returns the computed leaf hash of "MTH({d(0)}) = SHA-256(0x00 || d(0))".
+///
+/// # Arguments
+///
+/// * 'data' - The hash of the leaf data.
+pub fn leaf_digest(data: b256) -> b256 {
+    sha256((LEAF, data))
+}
+
+/// Returns the computed node hash of "MTH(D[n]) = SHA-256(0x01 || MTH(D[0:k]) || MTH(D[k:n]))".
+///
+/// # Arguments
+///
+/// * 'left' - The hash of the left node.
+/// * 'right' - The hash of the right node.
+pub fn node_digest(left: b256, right: b256) -> b256 {
+    sha256((NODE, left, right))
+}
+
+/// Calculates the length of the path to a leaf
+///
+/// # Arguments
+///
+/// * `key` - The key or index of the leaf.
+/// * `num_leaves` - The total number of leaves in the Merkle Tree.
+fn path_length_from_key(key: u64, num_leaves: u64) -> u64 {
+    let mut total_length = 0;
+    let mut num_leaves = num_leaves;
+    let mut key = key;
+
+    while true {
+        // The height of the left subtree is equal to the offset of the starting bit of the path
+        let path_length = starting_bit(num_leaves);
+        // Determine the number of leaves in the left subtree
+        let num_leaves_left_sub_tree = (1 <<(path_length - 1));
+
+        if key <= (num_leaves_left_sub_tree - 1) {
+            // If the leaf is in the left subtreee, path length is full height of the left subtree
+            total_length = total_length + path_length;
+            break;
+        } else if num_leaves_left_sub_tree == 1 {
+            // If the left sub tree has only one leaf, path has one additional step
+            total_length = total_length + 1;
+            break;
+        } else if (num_leaves - num_leaves_left_sub_tree) <= 1 {
```

Comment:
Did you mean this?
```suggestion
        } else if (num_leaves - num_leaves_left_sub_tree) == 1 {
```

Reply:
This is intentional and is follow [fuel-merkle-sol](https://github.com/FuelLabs/fuel-merkle-sol/blob/90e8463074d9c038ef7f0b15d3bb91549812bb02/contracts/tree/Utils.sol#L41)